casper/src/main/resources/PoS.rhox

```diff
@@ -155,26 +154,61 @@ new  PoS,
               }
             } |
 
-            contract PoS (@"chargeDeploy", @deployerId, @amount, returnCh) = {
-              new vaultCh, transferCh, userCh, processPayCh,
-                  depositCh,
-                  newPendingCh in {
-                getUser!(deployerId, *userCh) |
-                runMVar!(*stateCh, *processPayCh, *returnCh) |
-                for(@state, payResultCh <- processPayCh;
-                    @userPk <- userCh) {
+            new currentDeployerData in {
+              contract PoS (@"chargeDeploy", @deployerId, @amount, return) = {
+                new depositCh in {
                   deposit!(deployerId, amount, *depositCh) |
-
-                  distributeRewards!(
-                    amount,
-                    state.get("activeValidators"),
-                    state.get("allBonds"),
-                    state.get("pendingRewards"),
-                    *newPendingCh) |
-
-                  for(@depositResult <- depositCh;
-                      @newPending <- newPendingCh) {
-                    payResultCh!(state.set("pendingRewards", newPending), depositResult)
+                  for(@depositResult <- depositCh){
+                    match depositResult {
+                      (true, _) => {
+                        currentDeployerData!((deployerId, amount)) |
+                        return!(depositResult)
+                      }
+                      (_, errorMessage) => {
+                        return!((false, "Deploy payment failed: " ++ errorMessage))
+                      }
+                    }
+                  }
+                }
+              } |
+
+              contract PoS (@"refundDeploy", @refundAmount, return) = {
+                for(@(deployerId, initialPaymentAmount) <- currentDeployerData){
+                  new revAddressCh in {
+                    revAddressOps!("fromDeployerId", deployerId, *revAddressCh) |
+                    for(@deployerRevAddress <- revAddressCh){
+                      new transferCh in {
+                        @posVault!("transfer", deployerRevAddress, refundAmount, posAuthKey, *transferCh) |
```

Comment:
What prevents a normal user from just calling `PoS!("refundDeploy", 99999999, ...)`?

Reply:
There are a number of PoS methods that will have to be made private to the protocol somehow. This interface assumes we'll get around to that. In that case, refund amounts, deployer identity, and ensuring that each `chargeDeploy` is followed by a `refundDeploy`.

Reply:
Please add a comment of that nature then